### ISSUE TITLE
wrong hash used to load css file after build

### ISSUE THREAD

**Issue by simonporter007** [expressive-code/expressive-code#222]
Hello!

I must be having a config issue somewhere as I searched and didn't see an exact description of the problem I'm having, and presuming other folks are using the collapsible sections plugin okay. What's happening is when I run `astro dev`, all works fine, no drama there. When running `astro build` and loading it with preview, it's attempting to load a CSS file that doesn't exist (is not part of the _astro build output).

I've added some logging where I can to try and track it, and it seems to be due to the CSS vars changing when using collapsible sections plugin. They go from `ec-cs-` to `ec-collapsibleSections` during the build process. As far as I can see, there's a first run through where the hash for the CSS gets generated, and then another run through later on that runs _after_ this change and so generates a different hash (for a file that doesn't exist in _astro).

Here's my general setup, as I'm not sure where the issue resides (plugin? astro config? usage?):
- updated expressive-code and plugins to latest versions: 0.35.3 (but this was happening before when I was on 0.33.x)
- using astro version 4.11.5
- using the Code component inside mdx files
- using the ec.config.mjs file (soon as i remove `pluginCollapsibleSections()` it all works again, css loads correctly)
- Astro is in SSG/static mode

```
export default defineEcConfig({
  emitExternalStylesheet: true, // was trying this with false, but still didn't work
  defaultProps: {
    wrap: true,
    preserveIndent: true,
  },
  themes: ['dracula', 'github-light'],
  plugins: [pluginLineNumbers(), pluginCollapsibleSections()],
  frames: true,
  styleOverrides: {
    editorTabBorderRadius: '0px',
    borderRadius: '0px',
  },
});
```

```
export default defineConfig({
  prefetch: {
    prefetchAll: false,
  },
  integrations: [
    react(),
    tailwind({
      applyBaseStyles: false,
    }),
    sitemap(),
    robotsTxt(),
    icon({
      include: {
        solar: [
          'home-2-outline',
          'headphones-round-sound-linear',
          'sun-2-bold',
          'moon-stars-bold',
          'info-circle-broken',
          'close-circle-broken',
          'notebook-square-broken',
          'shield-warning-broken',
        ],
        'material-symbols': ['article-outline'],
      },
    }),
    astroExpressiveCode(),
    mdx(),
  ],
  site: SITE_URL,
  markdown: {
    syntaxHighlight: false,
    remarkPlugins: [remarkGemoji, remarkReadingTime],
    rehypePlugins: [
      rehypeAccessibleEmojis,
      rehypeMeta,
      rehypeAutolinkHeadings,
      rehypeSlug,
    ],
    remarkRehype: { footnoteLabel: 'Footnotes' },
    gfm: true,
  },
  vite: { // tried removing this block in case it was interefering, didn't work
    plugins: [
      FontaineTransform.vite({
        fallbacks: ['Roboto'],
        resolvePath: (id) => new URL(`./public${id}`, import.meta.url), // id is the font src value in the CSS
      }),
    ],
  },
});
```

In case it helps, when running build, I see:
> building client (vite)
> 13:26:01 [types] Added src/env.d.ts type declarations.
> 13:26:04 [vite] ✓ 176 modules transformed.
> 13:26:04 [vite] dist/_astro/ec.3zb7u.js                            2.31 kB
> 13:26:04 [vite] dist/_astro/ec.s6lpm.css                          20.69 kB │ gzip:  4.49 kB
> 13:26:04 [vite] dist/_astro/page.C6xYrkd2.js                       0.04 kB │ gzip:  0.06 kB
`s6lpm` is the css file with e.g. `--ec-cs-closedFontFml`

it then runs:
>  generating static routes
> 13:26:04 ▶ src/pages/posts/[slug].astro
which generates the html files for the posts, and this is where it inserts the `<link>` stylesheet tag with a different hash. I logged this out during the render function, and this is the css that contains the `--ec-collapsibleSections` instead. In this case `f9cy2` is the hash, and doesn't exist in _astro.

I'm sure I must be missing something, but for the life of me, I can't see what 😅

**Comment by hippotastic** [comment]
Hello Simon, thanks for the thorough report and your attempts to analyze and narrow down the issue. I've never encountered this myself in many projects and can't reproduce this behavior.

There must be something really strange going on for the collapsible sections CSS to be present, yet contain the non-shortened `collapsibleSections` string, although the plugin calls `cssVarReplacements.set('collapsibleSections', 'cs')` on its very first line of the plugin function, which should cause the shortened version to be used instead.

The only possible explanation for this that I can come up with is that you have multiple conflicting versions of the Expressive Code packages, most importantly the core package, in your project, although you already tried your best to bring everything to the latest version. Please try deleting all package manager lockfiles (e.g. `package-lock.json` or `pnpm-lock.yaml`) in your project, deleting the entire contents of any `node_modules` folders in your project, and then letting your package manager reinstall them again.

**Comment by simonporter007** [comment]
Hey @hippotastic, thanks so much for the quick reply! I cleared out node modules and lock files, but still the same behaviour. It's so strange!

I tried to come up with a minimal reproction, but kept it mostly to the layout of the site i'm working on, as I guess I'm doing something weird with the site in general for this not to work, so didn't want to remove it entirely. I was able to reproduce [with this stackblitz](https://stackblitz.com/edit/withastro-astro-nkrouh?file=src%2Fpages%2Fposts%2F%5Bslug%5D.astro). It'll start in pnpm dev mode, where all works fine.

If you then quit dev mode, and run pnpm build and then pnpm preview, you should see the wrong hash attempted to be loaded and so the styling is missing from the post. If you remove the collapsible plugin from the ec config, then repeat - the css loads okay. You'll need to click the "test post" link on homepage when it runs, as that's where the code snippet lives (similar to the layout on my site).

Is it something to do with the way I call `post.render()` to get back the `<Content />` from the slug page perhaps?

**Comment by hippotastic** [comment]
Thanks a lot for the reproduction!

Phew, that seems to be a tricky one - I can see the issue in your project, but I'm struggling to find the difference to my projects, which all work fine (also with the latest versions of Astro and Starlight, making heavy use of content collections).

The only thing I could identify so far is that the issue is limited to the `Code` component. Fenced code blocks in MDX seem to work fine. I hope that knowing this can at least unblock you for now until the cause is found.

I'll need to keep investigating and probably start with a plain and working fresh project with EC and keep inserting your code piece by piece until I see it failing.

This will unfortunately take a while as I currently have very limited time due to other work. Sorry that I can't offer a quick solution right away!

**Comment by simonporter007** [comment]
No drama at all! Thanks for taking a look. I think I only went with the code component so I could keep the code in a separate file and get linting/prettier on it, but I can definitely work around that using fenced blocks though.

Let me know if I can help any but no rush by any means. Cheers

**Comment by hippotastic** [comment]
> For what it's worth, I had the same problem. Adding `@expressive-code/core` to my devDependencies actually _solves_ this problem for me though.

This actually works! I just tested it in the StackBlitz example posted by @simonporter007 by first upgrading all EC packages to `^0.37.1` (which didn't help on its own), and then running `pnpm i -D @expressive-code/core` to add EC core to devDependencies. After this last step, the hash is fine.

The only remaining issue (although it doesn't affect the output in any way) is this warning appearing on the console during build:
> "loadShikiTheme" is imported from external module "@expressive-code/core" but never used in "node_modules/.pnpm/@expressive-code+plugin-frames@0.37.1/node_modules/@expressive-code/plugin-frames/dist/index.js", "node_modules/.pnpm/@expressive-code+plugin-text-markers@0.37.1/node_modules/@expressive-code/plugin-text-markers/dist/index.js", "node_modules/.pnpm/@expressive-code+plugin-shiki@0.37.1/node_modules/@expressive-code/plugin-shiki/dist/index.js" and "node_modules/.pnpm/expressive-code@0.37.1/node_modules/expressive-code/dist/index.js".

I have to admit that I'm baffled about what is going on here. When comparing the output of `pnpm why @expressive-code/core` before and after adding the devDependency, the versions are exactly the same. I don't understand why adding core as an explicit dependency changes anything, and I don't understand why this makes the warning above appear.

**Comment by hippotastic** [comment]
This is now fixed in version 0.38.1, which I just released to NPM. Thank you for the report and reproduction!
